### PR TITLE
feat: res\synapse\workspace firewall child resource added

### LINKED ISSUE #1189: [AVM Module Issue]:  Synapse - Firewall and Sparkpool
### Check for previous/existing GitHub issues

- [X] I have checked for previous/existing GitHub issues

### Issue Type?

Feature Request

### Module Name

avm/res/synapse/workspace

### (Optional) Module Name if not listed above

_No response_

### (Optional) Module Version

_No response_

### Description

The synapse module should support the creation of firewall rules and sparkpools.

### (Optional) Correlation Id

_No response_

### LINKED ISSUE #2372: [Feature Request]: Add Firewall Rules to `synapse/workspace`
### Description

I'd like to request that firewall rules be created alongside synapse workspace module. 

('Microsoft.Synapse/workspaces/firewallRules@2021-06-01')

### PR DESCRIPTION
**Pull request by elanzel**
## Description

Fixes #1189
fixes #2372

## Pipeline Reference

| Pipeline |
| -------- |
|       [avm.res.synapse.workspace](https://github.com/elanzel/bicep-registry-modules/actions/workflows/avm.res.synapse.workspace.yml)   |

## Type of Change

- [ ] Update to CI Environment or utilities (Non-module affecting changes)
- [x] Azure Verified Module updates:
  - [x] Bugfix containing backwards-compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [x] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

### CODE REVIEW

File: avm/res/synapse/workspace/firewall-rules/main.bicep [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+param startIpAddress string
+
+@description('Required. The end IP address of the firewall rule. Must be IPv4 format. Must be greater than or equal to startIpAddress.')
+param endIpAddress string
+
+resource workspace 'Microsoft.Synapse/workspaces@2021-06-01' existing = {
+  name: workspaceName
+}
+
+resource firewallrules 'Microsoft.Synapse/workspaces/firewallRules@2021-06-01' = {
```
Comment (krbar) [inline avm/res/synapse/workspace/firewall-rules/main.bicep:21]:
```suggestion
resource firewallRule 'Microsoft.Synapse/workspaces/firewallRules@2021-06-01' = {
```
I think this should be singular and camelCase (as per [specs](https://azure.github.io/Azure-Verified-Modules/specs/bicep/#id-bcpnfr8---category-composition---code-styling---lower-camelcasing))

File: avm/res/synapse/workspace/firewall-rules/main.bicep [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  name: name
+  parent: workspace
+  properties: {
+    startIpAddress: startIpAddress
+    endIpAddress: endIpAddress
+  }
+}
+
+@description('The name of the deployed firewall rule.')
+output name string = firewallrules.name
```
Comment (krbar) [inline avm/res/synapse/workspace/firewall-rules/main.bicep:31]:
```suggestion
output name string = firewallRule.name
```
Following renamed symbolic name of the resource.

File: avm/res/synapse/workspace/firewall-rules/main.bicep [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+@description('The resource ID of the deployed firewall rule.')
+output resourceId string = firewallrules.id
```
Comment (krbar) [inline avm/res/synapse/workspace/firewall-rules/main.bicep:34]:
```suggestion
output resourceId string = firewallRule.id
```
Following renamed symbolic name of the resource.

File: avm/res/synapse/workspace/main.bicep [hunk 4]
```diff
@@ -322,6 +325,19 @@ resource workspace_roleAssignments 'Microsoft.Authorization/roleAssignments@2022
   }
 ]
 
+// Firewall Rules
+module workspace_firewall_rules 'firewall-rules/main.bicep' = [
```
Comment (krbar) [inline avm/res/synapse/workspace/main.bicep:329]:
```suggestion
module workspace_firewallRules 'firewall-rules/main.bicep' = [
```
Please change to camelCase.

File: avm/res/synapse/workspace/tests/e2e/fwrules/main.test.bicep [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+      location: resourceLocation
+      defaultDataLakeStorageAccountResourceId: nestedDependencies.outputs.storageAccountResourceId
+      defaultDataLakeStorageFilesystem: nestedDependencies.outputs.storageContainerName
+      sqlAdministratorLogin: 'synwsadmin'
+      firewallRules: [
+        {
+          name: 'fwrule01'
+          endIpAddress: '87.14.134.20'
+          startIpAddress: '87.14.134.20'
+          workspaceName: '${namePrefix}${serviceShort}001'
```
Comment (krbar) [inline avm/res/synapse/workspace/tests/e2e/fwrules/main.test.bicep:63]: Is the `workspaceName` set intentionally set here? It is not necessary as the parent module sets a default value.

**Review by krbar — DISMISSED** [review]

**Comment by AlexanderSehr** [comment]
Hey @elanzel,
Is this PR something that is still being worked on, or should we close it? :)

**Comment by elanzel** [comment]
> Hey @elanzel, Is this PR something that is still being worked on, or should we close it? :)

Hi @AlexanderSehr I need to apply the changes suggested by Kris

**Comment by elanzel** [comment]
Hi @krbar , i applied the changes and re-tested the code. Can you merge?

File: avm/res/synapse/workspace/firewall-rules/main.bicep (hunk 1 above)
Comment (AlexanderSehr) [reply avm/res/synapse/workspace/firewall-rules/main.bicep:21]: Un-resolved as not addressed

File: avm/res/synapse/workspace/firewall-rules/main.bicep (hunk 3 above)
Comment (AlexanderSehr) [reply avm/res/synapse/workspace/firewall-rules/main.bicep:34]: Un-resolved as not addressed

File: avm/res/synapse/workspace/tests/e2e/fwrules/main.test.bicep [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        {
+          name: 'fwrule02'
+          endIpAddress: '87.14.134.22'
+          startIpAddress: '87.14.134.21'
+        }
+      ]
+    }
+    dependsOn: [
+      nestedDependencies
+    ]
```
Comment (AlexanderSehr) [inline avm/res/synapse/workspace/tests/e2e/fwrules/main.test.bicep:73]:
```suggestion
```

File: avm/res/synapse/workspace/main.bicep [hunk 7]
```diff
@@ -66,6 +66,9 @@ param preventDataExfiltration bool = false
 @description('Optional. Enable or Disable public network access to workspace.')
 param publicNetworkAccess string = 'Enabled'
 
+@description('Optional. List of firewall rules to be created in the workspace.')
+param firewallRules array = []
```
Comment (AlexanderSehr) [inline avm/res/synapse/workspace/main.bicep:70]: I'd propose changing this to 
```bicep
param firewallRules firewallRuleType[]?
```
and add a corresponding type
```bicep
type firewallRuleType = {
  @description('Required. The name of the firewall rule.')
  name: string
  
  @description('Required. The start IP address of the firewall rule. Must be IPv4 format.')
  startIpAddress: string
  
  @description('Required. The end IP address of the firewall rule. Must be IPv4 format. Must be greater than or equal to startIpAddress.')
  endIpAddress: string
}
```

And adjusting the usage of the type to
```bicep
module workspace_firewallRules 'firewall-rules/main.bicep' = [
  for (rule, index) in (firewallRules ?? []): {
```
to not have it complain about the nullable parameter 😉 

Once done, please refresh the readme & json

File: avm/res/synapse/workspace/firewall-rules/main.bicep (hunk 1 above)
Comment (elanzel) [reply avm/res/synapse/workspace/firewall-rules/main.bicep:21]: updated

File: avm/res/synapse/workspace/tests/e2e/fwrules/main.test.bicep (hunk 5 above)
Comment (elanzel) [reply avm/res/synapse/workspace/tests/e2e/fwrules/main.test.bicep]: removed

File: avm/res/synapse/workspace/main.bicep (hunk 7 above)
Comment (elanzel) [reply avm/res/synapse/workspace/main.bicep]: Hi @AlexanderSehr i've applied all the changes and action status is green

**Review by AlexanderSehr — APPROVED** [review]

**Comment by AlexanderSehr** [comment]
🌊

**Pull request merged**